core/src/main/java/org/elasticsearch/index/query/QueryParseContext.java

```diff
@@ -115,10 +115,11 @@ public QueryBuilder parseTopLevelQueryBuilder() {
         @SuppressWarnings("unchecked")
         Optional<QueryBuilder> result = (Optional<QueryBuilder>) indicesQueriesRegistry.lookup(queryName, parseFieldMatcher,
                 parser.getTokenLocation()).fromXContent(this);
-        if (parser.currentToken() == XContentParser.Token.END_OBJECT) {
-            // if we are at END_OBJECT, move to the next one...
-            parser.nextToken();
+        if (parser.currentToken() != XContentParser.Token.END_OBJECT) {
+            throw new ParsingException(parser.getTokenLocation(),
+                    "[" + queryName + "] malformed query, expected [END_OBJECT] but found [" + parser.currentToken() + "]");
```

Comment:
this new check is just best effort, but should help finding out when parsers don't properly exhaust their own object. I don't think the previous leniency was needed.